test/test_fluid.py

```diff
@@ -0,0 +1,149 @@
+    """Test that the grad(v) returns {0, I} for v={constant, r_xyz}."""
+    from mirgecom.fluid import velocity_gradient
+    actx = actx_factory()
+
+    nel_1d = 16
+
+    from meshmode.mesh.generation import generate_regular_rect_mesh
+
+    mesh = generate_regular_rect_mesh(
+        a=(1.0,) * dim, b=(2.0,) * dim, n=(nel_1d,) * dim
+    )
+
+    order = 3
+    discr = EagerDGDiscretization(actx, mesh, order=order)
+    nodes = thaw(actx, discr.nodes())
+    zeros = discr.zeros(actx)
+    ones = zeros + 1.0
+
+    mass = 1*ones
+    for i in range(mass_exp):
+        mass *= (mass + i)
+    energy = zeros + 2.5
+    velocity = vel_fac * nodes
+    mom = mass * velocity
+
+    q = join_conserved(dim, mass=mass, energy=energy, momentum=mom)
+    cv = split_conserved(dim, q)
+
+    grad_q = obj_array_vectorize(discr.grad, q)
+    grad_cv = split_conserved(dim, grad_q)
+
+    grad_v = velocity_gradient(discr, cv, grad_cv)
+
+    tol = 1e-12
+    exp_result = vel_fac * np.eye(dim) * ones
+    grad_v_err = [discr.norm(grad_v[i] - exp_result[i], np.inf)
+                  for i in range(dim)]
+
+    assert max(grad_v_err) < tol
+
+
+@pytest.mark.parametrize("dim", [1, 2, 3])
+def test_velocity_gradient_eoc(actx_factory, dim):
+    """Test that the velocity gradient converges at the proper rate."""
+    from mirgecom.fluid import velocity_gradient
+    actx = actx_factory()
+
+    order = 3
+
+    from pytools.convergence import EOCRecorder
+    eoc = EOCRecorder()
+
+    nel_1d_0 = 5
+    for hn1 in [1, 2, 3, 4]:
+
+        nel_1d = hn1 * (nel_1d_0 - 1) + 1
+        h = 1/(nel_1d-1)
+
+        from meshmode.mesh.generation import generate_regular_rect_mesh
+        mesh = generate_regular_rect_mesh(
+            a=(1.0,) * dim, b=(2.0,) * dim, n=(nel_1d,) * dim
+        )
+
+        discr = EagerDGDiscretization(actx, mesh, order=order)
+        nodes = thaw(actx, discr.nodes())
+        zeros = discr.zeros(actx)
+        energy = zeros + 2.5
+
+        mass = nodes[dim-1]*nodes[dim-1]
+        velocity = make_obj_array([actx.np.cos(nodes[i]) for i in range(dim)])
+        mom = mass*velocity
+
+        q = join_conserved(dim, mass=mass, energy=energy, momentum=mom)
+        cv = split_conserved(dim, q)
+
+        grad_q = obj_array_vectorize(discr.grad, q)
+        grad_cv = split_conserved(dim, grad_q)
+
+        grad_v = velocity_gradient(discr, cv, grad_cv)
+
+        def exact_grad_row(xdata, gdim, dim):
+            exact_grad_row = make_obj_array([zeros for _ in range(dim)])
+            exact_grad_row[gdim] = -actx.np.sin(xdata)
+            return exact_grad_row
+
+        comp_err = make_obj_array([
+            discr.norm(grad_v[i] - exact_grad_row(nodes[i], i, dim), np.inf)
+            for i in range(dim)])
+        err_max = comp_err.max()
+        eoc.add_data_point(h, err_max)
+
+    print(eoc)
+    assert (
+        eoc.order_estimate() >= order - 0.5
+        or eoc.max_error() < 1e-9
+    )
```

Comment:
I'm actually not sure it makes sense to do a convergence test for this. Because the velocity gradient is not a weak (DG) gradient. In fact, what you're computing is sometimes referred to as the "broken gradient"; that is, the gradient operator restricted to cell volumes.

I think rather than run this convergence test, let's just choose a suitable resolution and just compute the norm of the broken gradient with the exact cell-wise gradient. @inducer @lukeolson Do you agree or disagree?

Reply:
I agree.  I think the test you suggest already in #298.  I didn't get this test properly updated to reflect the changes in its underlying implementation.